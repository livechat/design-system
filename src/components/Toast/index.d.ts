export type ToastHorizontalPosition = 'left' | 'center' | 'right';
export type ToastVerticalPosition = 'top' | 'middle' | 'bottom';

export interface ToastProps {
    children?: string;
    className: string,
    id?: string;
    autoHideDuration?: number;
    horizontalPosition?: ToastHorizontalPosition;
    verticalPosition?: ToastVerticalPosition;
    fixed?: boolean;
    success?: boolean;
    warning?: boolean;
    error?: boolean;
    info?: boolean;
    onClose?(): void;
}
