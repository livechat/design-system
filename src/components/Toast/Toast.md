Toast is a small message that by default shows up in the top middle of the screen. It disappears on its own after a few seconds. It provides a feedback about an operation for user.

### Toast types
Design System provides 5 different types of toasts. Use it with any library you want, ie. Redux. 

#### Success toast

```js
<Toast variant='success' removable>Message sent!</Toast>
```
```js noeditor
<ComponentHtmlMarkup>
  <Toast variant='success' removable>Message sent!</Toast>
</ComponentHtmlMarkup>
```

#### Warning Toast

```js
<Toast variant='warning' removable>Check if everything is fine.</Toast>
```
```js noeditor
<ComponentHtmlMarkup>
  <Toast variant='warning' removable>Check if everything is fine.</Toast>
</ComponentHtmlMarkup>
```

#### Error Toast

```js
<Toast variant='error' removable>Message could not be sent.</Toast>
```
```js noeditor
<ComponentHtmlMarkup>
  <Toast variant='error' removable>Message could not be sent.</Toast>
</ComponentHtmlMarkup>
```

#### Info Toast

```js
<Toast variant='info' removable>Sending message...</Toast>
```
```js noeditor
<ComponentHtmlMarkup>
  <Toast variant='info' removable>Sending message...</Toast>
</ComponentHtmlMarkup>
```

#### Notification Toast

```js
<Toast removable>Notification message</Toast>
```
```js noeditor
<ComponentHtmlMarkup>
  <Toast removable>Notification message</Toast>
</ComponentHtmlMarkup>
```

#### Toast with action

```js
const actionHandler = () => alert('click');

<Toast variant='error' action={{handler: actionHandler, label: 'Action'}} removable>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit,
  sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
</Toast>
```
```js noeditor
const actionHandler = () => alert('click');
<ComponentHtmlMarkup>
  <Toast variant='error' action={{handler: actionHandler, label: 'Action'}} removable>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit,
    sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
  </Toast>
</ComponentHtmlMarkup>
```
