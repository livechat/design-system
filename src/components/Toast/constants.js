export const HORIZONTAL_POSITION = ['left', 'center', 'right'];

export const VERTICAL_POSITION = ['top', 'middle', 'bottom'];

export const ANIMATION_TIME = 100;
