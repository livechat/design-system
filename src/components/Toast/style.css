.toast {
  display: flex;
  align-items: center;
  width: 360px;
  max-width: 90%;
  padding: 10px 5px 10px 12px;
  font-size: 15px;
  background: #ffffff;
  font-weight: 600;
  font-style: normal;
  font-stretch: normal;
  line-height: 20px;
  letter-spacing: normal;
  color: rgba(66,77,87,1);
  box-shadow: 0 6px 30px -10px rgba(0, 0, 0, 0.3);
  border-radius: 4px;
}

.toast svg {
  display: flex;
}

.toast-icon, .toast-close {
  padding-right: 12px;
  align-self: baseline;
}

.toast-icon svg {
  fill: #4384f5;
  height: 20px;
  width: 20px;
}

.toast-close {
  margin-left: auto;
}

.toast-close svg {
  fill: #9c999d;
  height: 20px;
  width: 20px;
  cursor: pointer;
}

.toast--fixed {
  position: fixed;
  z-index: 5318008;
}

.toast--horizontal-left {
  left: 5px;
}

.toast--horizontal-center {
  left: 50%;
  transform: translateX(-50%);
}

.toast--horizontal-right {
  right: 5px;
}

.toast--vertical-top {
  top: 5px;
}

.toast--vertical-middle {
  top: 50%;
  transform: translateY(-50%);
}

.toast--vertical-bottom {
  bottom: 5px;
}

.toast--centered {
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
}

.toast--horizontal-right {
  right: 5px;
}

.toast--vertical-top {
  top: 5px;
}

.toast--vertical-middle {
  top: 50%;
  transform: translateY(-50%);
}

.toast--vertical-bottom {
  bottom: 5px;
}

.toast--centered {
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
}

.toast--success {
  background: #4bb678;
}

.toast--warning {
  background: #efa842;
}

.toast--error {
  background: #f4574c;
}

.toast--info {
  background: #5ca8f5;
}

.toast--success, .toast--warning, .toast--error, .toast--info {
  color: #ffffff;
}

.toast--success svg, .toast--warning svg, .toast--error svg, .toast--info svg {
  fill: currentColor;
}