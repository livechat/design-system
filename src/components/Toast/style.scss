@import '../../scss/variables/colors';

$toast-appear-animation-timeout: 200ms;
$toast-exit-animation-timeout: 200ms;

$base-class: 'toast';

.#{$base-class} {
  display: inline-flex;
  align-items: flex-start;
  box-sizing: border-box;
  min-width: 300px;
  max-width: 500px;
  padding: 10px 12px 10px 12px;
  font-size: 15px;
  background: #fff;
  font-weight: 600;
  font-style: normal;
  font-stretch: normal;
  line-height: 20px;
  letter-spacing: normal;
  box-shadow: 0 6px 30px -10px rgba(0, 0, 0, 0.3);
  border-radius: 4px;
  margin-bottom: 5px;

  &__content {
    color: $font-primary-color;
  }

  &__icon,
  &__close {
    padding-right: 12px;
    align-self: baseline;
  }

  &__icon svg {
    display: flex;
    fill: $primary-color;
    height: 20px;
    width: 20px;
  }

  &__actions {
    margin-left: auto;
    display: flex;
    align-items: center;
    height: 20px;

    &-close {
      margin-left: auto;

      svg {
        display: flex;
        fill: $font-tertiary-color;
        height: 16px;
        width: 16px;
        cursor: pointer;
      }
    }

    &-custom {
      background-color: transparent;
      outline: 0;
      border: 0;
      margin: 0 16px;
      color: $font-primary-color;
      cursor: pointer;
      font-weight: 600;
      padding: 0;
      line-height: 20px;
    }
  }

  &--success {
    background: $success-color;
  }

  &--warning {
    background: $warning-color;
  }

  &--error {
    background: $error-color;
  }

  &--info {
    background: $info-color;
  }

  &--success,
  &--warning,
  &--error,
  &--info {
    .#{$base-class}__content,
    .#{$base-class}__actions-custom {
      color: #fff;
    }

    svg {
      fill: #fff;
    }
  }

  &-wrapper {
    display: flex;
    align-items: center;
    z-index: 120;
    position: absolute;

    &--fixed {
      position: fixed;
    }

    &--block {
      position: static;
    }

    &--horizontal-center {
      left: calc(50% - 50%);
    }

    &--horizontal-left {
      left: 5px;
    }

    &--horizontal-right {
      right: 5px;
    }

    &--vertical-top {
      top: 5px;

      .toast--animation-appear {
        transform: translateY(-100%) !important;
      }

      .toast--animation-appear.toast--animation-appear-active {
        transform: translateY(0) !important;
        transition: transform $toast-appear-animation-timeout ease-in;
      }

      .toast--animation-exit {
        transform: translateY(0) !important;
      }

      .toast--animation-exit.toast--animation-exit-active {
        transform: translateY(-100%) !important;
        transition: transform $toast-exit-animation-timeout ease-in;
      }
    }

    &--vertical-bottom {
      bottom: 5px;

      .toast--animation-appear {
        transform: translateY(100%) !important;
      }

      .toast--animation-appear.toast--animation-appear-active {
        transform: translateY(0) !important;
        transition: transform $toast-appear-animation-timeout ease-in;
      }

      .toast--animation-exit {
        transform: translateY(0) !important;
      }

      .toast--animation-exit.toast--animation-exit-active {
        transform: translateY(100%) !important;
        transition: transform $toast-exit-animation-timeout ease-in;
      }
    }
  }
}
