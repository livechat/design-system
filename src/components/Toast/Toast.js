import React from 'react';
import PropTypes from 'prop-types';
import CloseIcon from 'react-material-icon-svg/dist/CloseIcon';
import classNames from 'classnames/bind';
import { ToastIcon } from './ToastIcon';
import { VARIANTS } from './constants';
import styles from './style.scss';

const cx = classNames.bind(styles);

const Toast = props => {
  const {
    children,
    className,
    variant,
    onClose,
    removable,
    toastId,
    ...restProps
  } = props;

  const componentClassNames = `
    ${cx({
      toast: true,
      [`toast--${variant}`]: VARIANTS.some(option => option === variant)
    })} ${className}
  `;

  return (
    <div {...restProps} className={componentClassNames}>
      <div className={styles.toast__icon}>
        <ToastIcon variant={variant} />
      </div>
      <div className={styles.toast__content}>{children}</div>
      {removable && (
        <div
          className={styles.toast__close}
          aria-label="Close toast"
          onClick={onClose}
        >
          <CloseIcon />
        </div>
      )}
    </div>
  );
};

Toast.propTypes = {
  children: PropTypes.node.isRequired,
  variant: PropTypes.oneOf(['success', 'warning', 'error', 'info']),
  onClose: PropTypes.func,
  removable: PropTypes.bool
};

export default Toast;
