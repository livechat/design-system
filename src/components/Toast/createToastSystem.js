import { INFO, WARNING, ERROR, SUCCESS } from './constants';

class ToastSystem {
  constructor(container = null) {
    if (container) {
      this.container = container;
    }
  }

  setContainer = container => {
    this.container = container;
  };

  addToast(opts) {
    if (!this.container) {
      return;
    }

    const defaults = {
      content: '',
      variant: null,
      autoHideDelayTime: null,
      removable: false
    };

    this.container.addToast({ ...defaults, ...opts });
  }

  default = opts => {
    const { variant, ...restOpts } = opts;
    return this.addToast(restOpts);
  };

  success = opts => this.addToast({ ...opts, variant: SUCCESS });

  info = opts => this.addToast({ ...opts, variant: INFO });

  warning = opts => this.addToast({ ...opts, variant: WARNING });

  error = opts => this.addToast({ ...opts, variant: ERROR });
}

const createToastSystem = () => new ToastSystem();

export default createToastSystem;
