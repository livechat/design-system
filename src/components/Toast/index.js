import Toast from './Toast';

export { default as ToastWrapper } from './ToastWrapper';

export default Toast;
