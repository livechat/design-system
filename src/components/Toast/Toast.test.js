import React from 'react';
import { mount } from 'enzyme';
import Toast from './Toast';

const closeFn = jest.fn();

describe('Toast', () => {
  it('should render success toast', () => {
    const component = mount(
      <Toast error success>
        Toast content
      </Toast>
    );

    expect(component).toMatchSnapshot();
  });

  it('on close call passed function', () => {
    const component = mount(<Toast onClose={closeFn}>Click me</Toast>);
    const closeEl = component.find('.toast-close');

    closeEl.simulate('click');
    expect(closeFn).toHaveBeenCalled();
  });
});
