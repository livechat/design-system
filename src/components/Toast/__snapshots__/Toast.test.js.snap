// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should render success toast 1`] = `
<Toast
  variant="success"
>
  <div
    className="toast toast--success"
  >
    <div
      className="toast__icon"
    >
      <ToastIcon
        variant="success"
      >
        <CheckCircleIcon>
          <svg
            className="material material-check-circle-icon undefined"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2c5.52 0 10 4.48 10 10s-4.48 10-10 10S2 17.52 2 12 6.48 2 12 2m-1 14.5l7-7-1.41-1.41L11 13.67l-3.09-3.08L6.5 12l4.5 4.5z"
            />
          </svg>
        </CheckCircleIcon>
      </ToastIcon>
    </div>
    <div
      className="toast__content"
    >
      Toast content
    </div>
  </div>
</Toast>
`;
